### PR TITLE
Refactor DateSelector component

### PR DESCRIPTION
**Pull request by vagnervst**
*To ease review effort, please review each commit separately*

## Context
In order to have a step separated calendar where the user first chooses a preset on the sidebar and then is able to choose a date (if the preset allows it), I changed the structure of the `DateSelector` component so that it is state independent. This way, I transfered all the states to `DateInput` instead, also deprecating `CalendarInput`.

## Checklist
- [x] Remove states from `DateSelector`
- [x] Add separate Sidebar component to be used on `DateSelector`
- [x] Refactor `DateSelector` so that it doesn't have default presets and let the user choose all of them
- [x] Substitute direct `react-dates` usage on `DateSelector` by our `Calendar`
- [x] Update preset logic so that the user can't change de date defined by it (to be changed later)
- [x] Refactor `DateInput` to use the state independent `DateSelector` (implementing on it the state logic)
- [x] Deprecate `CalendarInput` because now we can achieve the same behavior of it by using DateSelector's prop `showSidebar` as false

## Linked Issues
- [ ] Resolves https://github.com/pagarme/pilot/issues/710 (Pilot implementation needed)

### Preview:
feb-19-2019 09-53-39

## How to test
1. Link `former-kit-skin-pagarme` dependency branch
```
git checkout add/aside-component; yarn build; yarn link
```

2. Link this repository with `former-kit-skin-pagarme`
```
git checkout add/aside-component; yarn; yarn link 'former-kit-skin-pagarme'
```

3. Run storybook
```
yarn storybook
```

4. Test DateSelector, DateInput and CalendarInput.

## Deploy Notes
Depends on https://github.com/pagarme/former-kit-skin-pagarme/pull/163

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: e4de299e7be08dd30c4a7a786a1009198ec08da0
  * Always use the standard `scope: action description`, always in english  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: e4de299e7be08dd30c4a7a786a1009198ec08da0
  * Always use the standard `scope: action description`, always in english  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 8058fa05aa058b7a743751496effaf10ad9e28a8
  * Always use the standard `scope: action description`, always in english  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 8058fa05aa058b7a743751496effaf10ad9e28a8
  * Always use the standard `scope: action description`, always in english  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

File: src/DateInput/DateInput.test.js [hunk 1]
```diff
@@ -1,28 +1,25 @@
 import React from 'react'
 import { mount } from 'enzyme'
+import { cleanup, render, fireEvent, waitForElement } from 'react-testing-library'
```
Comment (alalbux) [inline src/DateInput/DateInput.test.js]: could you break lines here? Is better to refactor imports

File: src/DateSelector/Aside.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  renderRadio ({ date, key, label }) {
+    const {
+      name,
+      selectedPreset,
+    } = this.props
+
+    const id = `${name}${key}`
+
+    return (
+      <li key={`${key}${label}`}>
```
Comment (alalbux) [inline src/DateSelector/Aside.js]: could you create a variable like make with `const id`?
```suggestion
      <li key={key}>
```

File: src/DateSelector/Aside.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    return (
+      <div className={theme.aside}>
+        {this.renderPresets(addProp({ name }, presets))}
+      </div>
+    )
+  }
+}
+
+Aside.propTypes = {
```
Comment (alalbux) [inline src/DateSelector/Aside.js]: Could you describe the propTypes properties? This is help us with docs

File: src/CalendarInput/index.js [hunk 4]
```diff
@@ -45,6 +45,13 @@ const isPeriodSelection = equals(PERIOD_SELECTION)
  */
 class CalendarInput extends Component {
   constructor (props) {
+    if (process.env.NODE_ENV !== 'test') {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'CalendarInput component is being deprecated, use DateInput with showCaledar prop as "false" instead.'
```
Comment (miguelssrs) [inline src/CalendarInput/index.js]: typo!!11!!
```suggestion
        'CalendarInput component is being deprecated, use DateInput with showCalendar prop as "false" instead.'
```

File: src/Calendar/index.js [hunk 5]
```diff
@@ -96,32 +96,41 @@ const getDatesYear = ({ end, start }) => {
  * the `react-dates`.
  */
 class Calendar extends Component {
+  static getDerivedStateFromProps ({ dates, focusedInput }, state) {
+    /** TODO
+     * We should probably refactor this 'visibleMonths' state logic because
+     * it seems there are a lot of black magic going on here
+     */
+    if (state.visibleMonths) {
```
Comment (miguelssrs) [inline src/Calendar/index.js]: Maybe we have a problem in `state.visibleMonths`. When we update the currentMonth to another month in the past or the future, this weird behavior happens and nothing update:
feb-19-2019 16-59-13
And this message appears on console:
`Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.`

File: src/Calendar/index.js [hunk 6]
```diff
@@ -96,32 +96,41 @@ const getDatesYear = ({ end, start }) => {
...
+      const year = getDatesYear(dates)
+      const visibleDates = isVisibleDate(state.visibleMonths, dates, year)
+
+      if (needCalendarUpdate(focusedInput, visibleDates)) {
+        return {
+          focusedInput: focusedInput || state.focusedInput,
```
Comment (miguelssrs) [inline src/Calendar/index.js]: You could put this in a variable to avoid repetition here and in line 117.

File: src/CalendarInput/index.js [hunk 7]
```diff
@@ -45,6 +45,13 @@ const isPeriodSelection = equals(PERIOD_SELECTION)
...
+      )
+    }
+
     super(props)
     const { start, end } = props.value
     const isVaidStart = (!start && !isPeriodSelection(props.dateSelection))
```
Comment (miguelssrs) [inline src/CalendarInput/index.js]: typooooo!!1!!! but I recommend fixing this in line 60 too.
```suggestion
    const isValidStart = (!start && !isPeriodSelection(props.dateSelection))
```

File: src/Calendar/index.js (hunk 6 above)
Comment (vagnervst) [reply src/Calendar/index.js]: Most of the logic on this function was removed, so this refactor is not necessary anymore 🙂

File: src/Calendar/index.js (hunk 5 above)
Comment (vagnervst) [reply src/Calendar/index.js]: Fixed! I debugged along with @Deivis and this bug was a result of a hotfix we needed on the past, now we solved this by removing most of the logic based on `visibleMonths` that updated the `inputKey` state.

File: src/Calendar/index.js [hunk 8]
```diff
@@ -96,32 +96,41 @@ const getDatesYear = ({ end, start }) => {
  * the `react-dates`.
  */
 class Calendar extends Component {
+  static getDerivedStateFromProps ({ dates, focusedInput }, state) {
+    /** TODO
```
[the remark this answers is not in the document]
Comment (vagnervst) [reply src/Calendar/index.js]: I believe it was an important alert to be maintained, but we removed this code already as it was causing the infinite months navigation bug you reported below 🙂

File: src/Calendar/index.js [hunk 9]
```diff
@@ -138,17 +90,22 @@ class Calendar extends Component {
...
+
+    if (onFocusChange) {
+      onFocusChange(focusedInput)
     }
   }
 
   handleMonthChange (firstDay, callback) {
     this.setState({
-      visibleMonths: this.getVisibleMonths(firstDay),
+      visibleMonths: this.getVisibleMonths(firstDay), // eslint-disable-line react/no-unused-state
```
Comment (Deivis) [inline src/Calendar/index.js]: This `visibleMonths`  prop isn't been used anymore, we should remove it, and refactor it usage in the functions which use this like `handleDatesChange `

Comment (vagnervst) [reply src/Calendar/index.js]: Good! I've done it and now the component is a lot more simplified.

File: src/DateSelector/Aside.js (hunk 3 above)
Comment (vagnervst) [reply src/DateSelector/Aside.js]: Done!

File: src/Calendar/index.js [hunk 10]
```diff
@@ -48,78 +45,31 @@ const getExtraProps = omit([
...
-
-  return moment().year()
-}
 /**
  * Custom calendar based on `react-dates` from airbnb with a simple interface.
  * This component have some special props and the surplus props are passed down to
  * the `react-dates`.
  */
 class Calendar extends Component {
+  static getDerivedStateFromProps ({ focusedInput }, state) {
```
Comment (rsmelo) [inline src/Calendar/index.js]: maybe you can use `componentDidUpdate` React docs https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops, suggest to use `getDerivedStateFromProps`in rare cases. This code is always updating state, but you just need to update the state if the new `prop.focusedInput` is different from previous value. Based on this rule `componentDidUpdate` can cover this case.

example:

```js
  componentDidUpdate (prevProps) {
    const { focusedInput } = this.props
    if (prevProps.focusedInput !== focusedInput) {
      // eslint-disable-next-line
      this.setState({
        focusedInput,
      })
    }
  }
```

Comment (vagnervst) [reply src/Calendar/index.js]: Good idea, this way we can simplify the component structure. I did it but changing a little the way I compared both props and states 🙂

File: stories/DateSelector/aside.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import Aside from '../../src/DateSelector/Aside'
+import presets from './presets'
+
+/*
+  TODO
+  Storybook action causes a 'Object(...) is not a function' exception,
+  change this when this issue (https://github.com/storybooks/storybook/issues/4796) is solved
+*/
+const action = (func, params) => {
+  console.log(func, params) // eslint-disable-line no-console
```
Comment (alalbux) [inline stories/DateSelector/aside.js]: Could you use `addons-actions` instead of `console.log`? I think that's better to visualization in storybook

Comment (Deivis) [reply stories/DateSelector/aside.js]: Nope, this will trigger a problem in comment above ☝️

Comment (Deivis) [reply stories/DateSelector/aside.js]: Correction, my fault, this issue was solved in `@storybook/react 4.1.x` and we are using `@storybook/react 4.1.7`.  So this action function should be replaced by Storybook action function.

Comment (alalbux) [reply stories/DateSelector/aside.js]: @Deivis in the issue has this comment https://github.com/storybooks/storybook/issues/1153#issuecomment-442165039 and you can use like that `action('test-action')('one')`. I test and works

Comment (uselessdev) [reply stories/DateSelector/aside.js]: LOL its a curry 😂

Comment (alalbux) [reply stories/DateSelector/aside.js]: @vagnervst it`s necessary do this in all files that you create in the stories scope `action` function

File: src/CalendarInput/index.js [hunk 12]
```diff
@@ -45,15 +45,23 @@ const isPeriodSelection = equals(PERIOD_SELECTION)
  */
 class CalendarInput extends Component {
   constructor (props) {
+    if (process.env.NODE_ENV !== 'test') {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'CalendarInput component is being deprecated, use DateInput with showCalendar prop as "false" instead.'
+      )
```
Comment (uselessdev) [inline src/CalendarInput/index.js]: Please, can you add an `@deprecated` with the same message above the class declaration?

File: src/DateSelector/index.js [hunk 13]
```diff
@@ -86,248 +143,150 @@ class DateSelector extends Component {
...
+      selectedPreset,
+      onChange,
+      presets,
+    } = this.props
 
-      const foundPreset = findItemByKey(presets)
+    const preset = getPreset(selectedPreset, presets)
 
-      this.handlePresetChange(foundPreset.date(), selectedPreset)
+    if (!selectedPreset || preset.date() === null) {
```
Comment (Deivis) [inline src/DateSelector/index.js]: You already have `ramda` on this page, why not use an `isNil`?
```suggestion
    if (!selectedPreset || isNil(preset.date())) {
```

File: src/DateSelector/index.js [hunk 14]
```diff
@@ -86,248 +143,150 @@ class DateSelector extends Component {
...
-              endDate={end}
-              onDatesChange={this.handleDatesChange}
-              hideKeyboardShortcutsPanel
-            />
-          )
-        }
+        <Calendar
+          numberOfMonths={2}
+          daySize={40}
+          isDayBlocked={date => presetRange !== null && !isDateBetweenRange(date, { start, end })}
```
Comment (Deivis) [inline src/DateSelector/index.js]: This logic could be in a function outside the component just to keep more readable this chunk of code.

File: src/DateSelector/DateSelector.test.js [hunk 15]
```diff
@@ -23,318 +43,221 @@ const defaultDates = {
...
     )
 
-    component.find('input').at(1).simulate('change')
-    expect(onChange).toHaveBeenCalledWith({ start: null, end: null })
+    fireEvent.click(container.querySelector('#open-button'))
+
+    await waitForPopover(container)
+
+    const now = moment()
+    fireEvent.click(container.querySelector(`td[aria-label="${now.format('dddd, MMMM D, YYYY')}"]`))
```
Comment (Deivis) [inline src/DateSelector/DateSelector.test.js]: Too long
```suggestion
    fireEvent.click(
       container
       .querySelector(`td[aria-label="${now.format('dddd, MMMM D, YYYY')}"]`)
    )
```

File: src/Calendar/index.js [hunk 16]
```diff
@@ -48,48 +45,11 @@ const getExtraProps = omit([
...
-  focusedInput,
-  {
-    isVisibleStart,
-    isVisibleEnd,
-  }
-) => (!isVisibleEnd || (focusedInput === START_DATE && !isVisibleStart))
 
 const isOutsideRange = limits => complement(validateDate(limits))
 
 const validateVisibleMonths = (currentMonth, months) => times(add(currentMonth), months)
```
Comment (Deivis) [inline src/Calendar/index.js]: Too long 😄 
```suggestion
const validateVisibleMonths = (currentMonth, months) =>
  times(add(currentMonth), months)
```

File: src/CalendarInput/index.js (hunk 12 above)
Comment (vagnervst) [reply src/CalendarInput/index.js]: I added the deprecated flag [here](https://github.com/pagarme/former-kit/pull/228/files#diff-df9b7fa81b80bb7edfe2325fb597d3a5R37), is it correct?

File: stories/DateSelector/aside.js (hunk 11 above)
Comment (vagnervst) [reply stories/DateSelector/aside.js]: Great! Used `action` from `@storybook/addon-actions` and it worked like a charm.

**Comment by vagnervst** [comment]
Added some fixes to MaskedInput usage, handling `state.dates` of DateInput as text so that every change made by the user is reflected on the text fields, a behavior I couldn't achieve when `state.dates` were moment objects.

File: src/CalendarInput/index.js [hunk 17]
```diff
@@ -34,6 +34,7 @@ const SINGLE_SELECTION = 'single'
 const isPeriodSelection = equals(PERIOD_SELECTION)
 
 /**
+ * @deprecated
  * This component is a simplified version of the date selector only with the basic
  * and the calendar inputs.
  * The input component designed to receive one or two dates, have a date format mask
```
Comment (uselessdev) [inline src/CalendarInput/index.js]: @vagnervst add at the end of component description.

```suggestion
 * The input component designed to receive one or two dates, have a date format mask
 *
 * @deprecated CalendarInput component is being deprecated, use DateInput with showCalendar prop as "false" instead.
```

Comment (vagnervst) [reply src/CalendarInput/index.js]: Done!

File: stories/DateInput/index.js [hunk 18]
```diff
@@ -1,105 +1,123 @@
...
+    </Section>
+  ))
+  .add('with sidebar and selected preset', () => (
+    <Section>
+      <DateInputState
+        selectedPreset="last-7"
+      />
+    </Section>
+  ))
+  .add('with sidebar only', () => (
```
Comment (alalbux) [inline stories/DateInput/index.js]: could you change the title off story to "with sideber only first" or some thing like that? Cause the calendar opens after select a date

**Pull request merged**